pkg/edvprovider/couchdbedvprovider/couchdbedvprovider.go

```diff
@@ -0,0 +1,247 @@
+	}
+
+	return c.coreStore.CreateIndex(createIndexRequest)
+}
+
+// Get fetches the document associated with the given key.
+func (c CouchDBEDVStore) Get(k string) ([]byte, error) {
+	return c.coreStore.Get(k)
+}
+
+// CreateEDVIndex creates the index which will allow for encrypted indices to work.
+func (c CouchDBEDVStore) CreateEDVIndex() error {
+	createIndexRequest := storage.CreateIndexRequest{
+		IndexStorageLocation: "EDV_EncryptedIndexesDesignDoc",
+		IndexName:            "EDV_IndexName",
+		WhatToIndex:          `{"fields": ["` + mapDocumentIndexedField + `"]}`,
+	}
+
+	return c.coreStore.CreateIndex(createIndexRequest)
+}
+
+// Query does an EDV encrypted index query.
+// We first get the "mapping document" and then use the ID we get from that to lookup the associated encrypted document.
+// Then we check that encrypted document to see if the value matches what was specified in the query.
+func (c CouchDBEDVStore) Query(query *models.Query) ([]string, error) {
+	idsOfDocsWithMatchingQueryIndexName, err := c.findDocsMatchingQueryIndexName(query.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.filterDocsByQuery(idsOfDocsWithMatchingQueryIndexName, query)
+}
+
+func (c CouchDBEDVStore) findDocsMatchingQueryIndexName(queryIndexName string) (map[string]struct{}, error) {
+	itr, err := c.coreStore.Query(`{
+		   "selector": {
+		       "` + mapDocumentIndexedField + `": "` + queryIndexName + `"
+		   },
+			"use_index": ["EDV_EncryptedIndexesDesignDoc", "EDV_IndexName"]
+		}`)
+	if err != nil {
+		return nil, err
+	}
+
+	ok, err := itr.Next()
+	if err != nil {
+		return nil, err
+	}
+
+	idsOfDocsWithAMatchingIndex := make(map[string]struct{})
+
+	for ok {
+		value, valueErr := itr.Value()
+		if valueErr != nil {
+			return nil, valueErr
+		}
+
+		receivedCouchDBIndexMappingDocument := couchDBIndexMappingDocument{}
+
+		err = json.Unmarshal(value, &receivedCouchDBIndexMappingDocument)
+		if err != nil {
+			return nil, err
+		}
+
+		idsOfDocsWithAMatchingIndex[receivedCouchDBIndexMappingDocument.MatchingEncryptedDocID] = struct{}{}
+
+		ok, err = itr.Next()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	err = itr.Release()
+	if err != nil {
+		return nil, err
+	}
+
+	return idsOfDocsWithAMatchingIndex, nil
+}
+
+// Given a set of documents, returns the document IDs that satisfy the query.
+func (c CouchDBEDVStore) filterDocsByQuery(docIDs map[string]struct{}, query *models.Query) ([]string, error) {
+	matchingDocIDs := make([]string, 0)
+
+	for docID := range docIDs {
+		documentBytes, err := c.coreStore.Get(docID)
+		if err != nil {
+			if err == storage.ErrValueNotFound {
+				return nil, edverrors.ErrDocumentNotFound
+			}
+
+			return nil, err
+		}
+
+		foundEncryptedDoc := models.EncryptedDocument{}
+
+		err = json.Unmarshal(documentBytes, &foundEncryptedDoc)
+		if err != nil {
+			return nil, err
+		}
+
+		if documentMatchesQuery(foundEncryptedDoc, query) {
```

Comment:
Is this necessary? i.e. will there ever be a document returned by the mapping-based query that wouldn't match?

Reply:
Yes and yes - at this point in the function, the documents that have been identified are those that have a matching index name, however the actual value associated with the index may not match. `documentMatchesQuery` checks to ensure the value matches as well.